File: src/main/java/yodelr/Yodelr.java
/**
 * Copyright Vespa.ai.
 * Not for redistribution outside of candidate interview context.
 */

package yodelr;

import java.util.List;

/**
 * The Yodelr service interface.
 *
 * This allows adding and deleting users, adding and retrieving posts
 * and getting trending topics.
 */
public interface Yodelr {

    void addUser(String userName);

    void addPost(String userName, String postText, long timestamp);

    void deleteUser(String userName);

    List<String> getPostsForUser(String userName);

    List<String> getPostsForTopic(String topic);

    List<String> getTrendingTopics(long fromTimestamp, long toTimestamp);

    static Yodelr make() { return new YodelrImpl(); }

}

File: assignment/yodelr/yodelr.py
# Copyright Vespa.ai.
# Not for redistribution outside of candidate interview context.

from abc import ABC, abstractmethod
from typing import List


class Yodelr(ABC):
    """
    The Yodelr service interface.

    This allows adding and deleting users, adding and retrieving posts
    and getting trending topics.
    """

    @abstractmethod
    def add_user(self, user_name: str) -> None:
        pass

    @abstractmethod
    def add_post(self, user_name: str, post_text: str, timestamp: int) -> None:
        pass

    @abstractmethod
    def delete_user(self, user_name: str) -> None:
        pass

    @abstractmethod
    def get_posts_for_user(self, user_name: str) -> List[str]:
        pass

    @abstractmethod
    def get_posts_for_topic(self, topic: str) -> List[str]:
        pass

    @abstractmethod
    def get_trending_topics(self, from_timestamp: int, to_timestamp: int) -> List[str]:
        pass
